src/geom/SphPoint.cc

```diff
@@ -0,0 +1,244 @@
+                       sin(_longitude)*cos(_latitude),
+                                       sin(_latitude));
+    }
+
+    Angle SphPoint::operator[](size_t index) const {
+        switch (index) {
+          case 0:
+            return getLongitude();
+          case 1:
+            return getLatitude();
+          default:
+            throw pexExcept::OutOfRangeError("Index " + to_string(index) + " must be 0 or 1.");
+        }
+    }
+
+    bool SphPoint::operator==(SphPoint const & other) const noexcept {
+        if (abs(_latitude) < HALFPI) {
+            return _longitude == other._longitude && _latitude == other._latitude;
+        } else {
+            return _latitude == other._latitude;
+        }
+    }
+
+    bool SphPoint::operator!=(SphPoint const & other) const noexcept {
+        return !(*this == other);
+    }
+
+    Angle SphPoint::bearingTo(SphPoint const & other) const {
+        if (abs(_latitude) >= HALFPI) {
+            stringstream buffer;
+            buffer << "Cannot calculate offset from pole " << *this << ".";
+            throw pexExcept::DomainError(buffer.str());
+        }
+
+        Angle const deltaLon = other.getLongitude()- this->getLongitude();
+
+        double const sinDelta1 = sin(      getLatitude().asRadians());
+        double const sinDelta2 = sin(other.getLatitude().asRadians());
+        double const cosDelta1 = cos(      getLatitude().asRadians());
+        double const cosDelta2 = cos(other.getLatitude().asRadians());
+
+        // Adapted from http://www.movable-type.co.uk/scripts/latlong.html
+        double const y = sin(deltaLon)*cosDelta2;
+        double const x = cosDelta1*sinDelta2 - sinDelta1*cosDelta2*cos(deltaLon);
+        return wrap(90.0*degrees - atan2(y, x)*radians);
+    }
+
+    Angle SphPoint::separationTo(SphPoint const & other) const noexcept {
+        return haversine(getLongitude() - other.getLongitude(),
+                         getLatitude()  - other.getLatitude(),
+                         cos(getLatitude().asRadians()),
+                         cos(other.getLatitude().asRadians())
+        );
+    }
+
+    SphPoint SphPoint::rotated(SphPoint const & axis, Angle const & amount) const {
+        double const theta = amount.asRadians();
+        if (!isfinite(theta)) {
+            stringstream buffer;
+            buffer << "NaN rotation angle " << amount << ".";
+            throw pexExcept::InvalidParameterError(buffer.str());
+        }
+
+        auto const rotation = Eigen::AngleAxisd(theta, axis.getVector().asEigen()).matrix();
+        auto const x = getVector().asEigen();
+        auto const xprime = rotation * x;
+        return SphPoint(Point3D(xprime));
+    }
+
+    SphPoint SphPoint::offset(Angle const & bearing, Angle const & amount) const {
+        double const phi = bearing.asRadians();
+        if (!isfinite(phi)) {
+            stringstream buffer;
+            buffer << "NaN offset bearing " << bearing << ".";
```

Comment:
I suggest you simply propagate NaNs so as to avoid unnecessarily raising exceptions.